plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -506,13 +508,53 @@ public FlowExecutionOwner getOwner() {
     private TimingFlowNodeStorage createStorage() throws IOException {
         FlowNodeStorage wrappedStorage;
 
-        FlowDurabilityHint hint = getDurabilityHint();
-        wrappedStorage = (hint.isPersistWithEveryStep())
-                ? new SimpleXStreamFlowNodeStorage(this, getStorageDir())
-                : new BulkFlowNodeStorage(this, getStorageDir());
+        if (this.storageDir != null && this.storageDir.endsWith("-completed")) {
+           wrappedStorage = new BulkFlowNodeStorage(this, getStorageDir());
+        } else {
+            FlowDurabilityHint hint = getDurabilityHint();
+            wrappedStorage = (hint.isPersistWithEveryStep())
+                    ? new SimpleXStreamFlowNodeStorage(this, getStorageDir())
+                    : new BulkFlowNodeStorage(this, getStorageDir());
+        }
         return new TimingFlowNodeStorage(wrappedStorage);
     }
 
+    /**
+     * Called when the build completes to migrate from {@link SimpleXStreamFlowNodeStorage} to
+     * {@link BulkFlowNodeStorage} to improve read performance for completed builds.
+     */
+    private void optimizeStorage(FlowNode flowEndNode) {
+        if (storage.delegate instanceof SimpleXStreamFlowNodeStorage) {
+            String newStorageDir = (this.storageDir != null) ? this.storageDir + "-completed" : "workflow-completed";
+            try {
+                FlowNodeStorage newStorage = new BulkFlowNodeStorage(this, new File(this.owner.getRootDir(), newStorageDir));
+                DepthFirstScanner scanner = new DepthFirstScanner();
+                scanner.setup(flowEndNode);
+                // The hope is that by doing this right when the execution completes, most of the nodes will already be
+                // cached in memory, saving us the cost of having to read them all from disk.
+                for (FlowNode node : scanner) {
+                    newStorage.storeNode(node, true);
+                }
+                newStorage.flush();
+                File oldStorageDir = getStorageDir();
+                synchronized (this) {
+                    this.storageDir = newStorageDir;
+                    // TODO: A more conservative option could be to instead keep using the current storage until after
+                    // a restart. In that case we'd have to defer deletion of the old storage dir and handle it in
+                    // initializeStorage.
+                    this.storage = new TimingFlowNodeStorage(newStorage);
```

Comment:
Does this require a `build.xml` save? Or is that done by something else anyway?

Reply:
The two places that call `onProgramEnd`  (`CpsFlowExecution.croak` and `CpsThreadGroup.run`) both call `saveOwner` right after the call, so I think things will be ok, but I will check in more detail later.

Reply:
One change I will make is to expand the outer catch clause in `optimizeStorage` to catch `Exception` in case `DepthFirstScanner` can throw an NPE or similar runtime exceptions in unusual cases.

Other than that, I think things should be ok unless we ever see a `Throwable` that is not an `Exception` in `onProgramEnd` (e.g. `StackOverflowError`, `OutOfMemoryError`), but that is the same with or without this PR. I think we could perhaps catch `Throwable` or adapt existing callers to use `try`/`finally` to be more robust against thrown `Error`s but I am not sure that it would make a difference in practice.